source/glbinding/source/RingBuffer.hpp

```diff
@@ -0,0 +1,243 @@
+{
+    sizeType tail = m_tails[key].load(std::memory_order_acquire);
+    if (tail == m_head.load(std::memory_order_acquire))
+    {
+        return it;
+    }
+
+    sizeType nextTail = next(tail);
+    m_tails[key].store(nextTail, std::memory_order_release);
+    updateTail();
+    return cbegin(key);
+}
+
+template <typename T>
+void RingBuffer<T>::release(TailIdentifier key, const typename std::vector<T>::const_iterator & it)
+{
+    sizeType newTail = std::distance(m_buffer.begin(), it);
+    m_tails[key].store(newTail, std::memory_order_release);
+    updateTail();    
+}
+
+template <typename T>
+typename RingBuffer<T>::sizeType RingBuffer<T>::size(TailIdentifier key)
+{
+    sizeType head = m_head.load(std::memory_order_acquire);
+    sizeType tail = m_tails[key].load(std::memory_order_acquire);
+    return size(head, tail);
+}
+
+template <typename T>
+typename RingBuffer<T>::sizeType RingBuffer<T>::maxSize() {
+    return m_size - 1;
+}
+
+template <typename T>
+typename RingBuffer<T>::sizeType RingBuffer<T>::size()
+{
+    sizeType head = m_head.load(std::memory_order_acquire);
+    sizeType tail = m_tail.load(std::memory_order_acquire);
+    return size(head, tail);
+}
+
+template <typename T>
+bool RingBuffer<T>::isFull()
+{
+    sizeType head = m_head.load(std::memory_order_relaxed);
+    sizeType nextHead = next(head);
+    if (nextHead == m_tail.load(std::memory_order_acquire)) {
+        return true;
+    }
+    return false;
+}
+
+template <typename T>
+bool RingBuffer<T>::isEmpty()
+{
+    sizeType tail = m_tail.load(std::memory_order_relaxed);
+    if (tail == m_head.load(std::memory_order_acquire)) {
+        return true;
+    }
+    return false;
+}
+
+
+//protected
+
+template <typename T>
+typename RingBuffer<T>::sizeType RingBuffer<T>::next(RingBuffer::sizeType current) {
+        return (current + 1) % m_size;
+}
+
+template <typename T>
+void RingBuffer<T>::updateTail()
+{
+    m_tail_mutex.lock();
+    sizeType tail = m_tail.load(std::memory_order_acquire);
+    sizeType currentMin = 2 * m_size;
+
+    for(auto it = m_tails.cbegin(); it != m_tails.cend(); ++it ) {
+        sizeType tailPos = it->second.load(std::memory_order_acquire);
+        if (tailPos == tail)
+        {
+            m_tail_mutex.unlock();
+            return;
+        }
+
+        if (tailPos < tail)
+        {
+            tailPos += m_size;
+        }
+
+        if (tailPos < currentMin)
+        {
+            currentMin = tailPos;
+        }
+    }
+
+    if (currentMin == 2 * m_size)
+    {
+        m_tail_mutex.unlock();
+        return;
+    }
+
+    if (currentMin >= m_size)
+    {
+        currentMin = currentMin % m_size;
+    }
+
+    m_tail.store(currentMin, std::memory_order_release);
+    m_tail_mutex.unlock();
+}
+
+template <typename T>
+typename RingBuffer<T>::sizeType RingBuffer<T>::size(sizeType head, sizeType tail)
+{
+    if (head < tail)
```

Comment:
I think this can be rewritten as ```(m_size + head - tail) % m_size``` but I don't know which solution I would prefer.